### PR TITLE
try out replacing Dsymbol.error() part 5

### PR DESCRIPTION
**Pull request by WalterBright**
more of https://github.com/dlang/dmd/pull/15657

### CODE REVIEW

File: compiler/test/unit/frontend.d [hunk 1]
```diff
@@ -388,5 +388,5 @@ unittest
     t.module_.fullSemantic();
 
     assert(global.errors == 1);
-    assert(diagnosticMessages[0] == "is a Ddoc file, cannot import it");
+    assert(diagnosticMessages[length - 32 .. length] == "is a Ddoc file, cannot import it");
```
Comment (thewilsonator) [inline compiler/test/unit/frontend.d:391]:
```
test/unit/frontend.d(391): Error: undefined identifier `length`
```

Comment (WalterBright) [reply compiler/test/unit/frontend.d]: thanks, Nik

File: compiler/test/unit/frontend.d [hunk 2]
```diff
@@ -388,5 +388,5 @@ unittest
     t.module_.fullSemantic();
 
     assert(global.errors == 1);
-    assert(diagnosticMessages[0] == "is a Ddoc file, cannot import it");
+    assert(diagnosticMessages[$ - 32 .. $] == "is a Ddoc file, cannot import it");
```
Comment (thewilsonator) [inline compiler/test/unit/frontend.d:391]: You want `diagnosticMessages[0][$ - 32 .. $]` there. Also I'd do it as 
```d
string msg = "is a Ddoc file, cannot import it";
assert(diagnosticMessages[0][$ - msg.length] == msg);
```
as it is more clear.

Comment (WalterBright) [reply compiler/test/unit/frontend.d]: You're right.

File: compiler/test/unit/frontend.d [hunk 3]
```diff
@@ -388,5 +388,6 @@ unittest
     t.module_.fullSemantic();
 
     assert(global.errors == 1);
-    assert(diagnosticMessages[0] == "is a Ddoc file, cannot import it");
+    string msg = "is a Ddoc file, cannot import it";
+    assert(diagnosticMessages[0][$ - 32 .. $] == msg);
```
Comment (thewilsonator) [inline compiler/test/unit/frontend.d:392]: `[$ - msg.length .. $]`

Comment (WalterBright) [reply compiler/test/unit/frontend.d]: I really should stop trying to do something else while I work on this.

**Review by thewilsonator — APPROVED** [review]

File: compiler/src/dmd/dmodule.d [hunk 4]
```diff
@@ -1525,7 +1528,7 @@ private const(char)[] processSource (const(ubyte)[] src, Module mod)
 
         if (buf.length & 1)
         {
-            mod.error("odd length of UTF-16 char source %llu", cast(ulong) buf.length);
+            .error(mod.loc, "%s `%s` odd length of UTF-16 char source %llu", mod.kind, mod.toPrettyChars, cast(ulong) buf.length);
```
Comment (ibuclaw) [inline compiler/src/dmd/dmodule.d:1531]: This caused a diagnostic regression - here, `mod.loc is Loc.initial`.

Before:
```
bom_error_UTF16.d: Error: module `bom_error_UTF16` odd length of UTF-16 char source 159
```

After:
```
Error: module `bom_error_UTF16` odd length of UTF-16 char source 159
```

**Pull request merged**